src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp

```diff
@@ -0,0 +1,1420 @@
+                                                      fluid, solid, localMatrix, localRhs, kernelFlags );
+      AccumulationKernel< NUM_EQN, NUM_DOF >::template launch< POLICY >( subRegion.size(), kernel );
+    } );
+  }
+
+};
+
+
+
+/******************************** PhaseMobilityKernel ********************************/
+
+/**
+ * @class PhaseMobilityKernel
+ * @tparam NUM_PHASE number of fluid phases
+ * @brief Define the interface for the property kernel in charge of computing the phase mobilities
+ */
+template< integer NUM_PHASE >
+class PhaseMobilityKernel
+{
+public:
+
+  //using Base = MultiphaseFluidAccessors::PropertyKernelBase< NUM_COMP >;
+
+  /// Compile time value for the number of phases
+  static constexpr integer numPhase = NUM_PHASE;
+
+  /**
+   * @brief Constructor
+   * @param[in] subRegion the element subregion
+   * @param[in] fluid the fluid model
+   * @param[in] relperm the relperm model
+   */
+  PhaseMobilityKernel( ObjectManagerBase & subRegion,
+                       TwoPhaseFluid const & fluid,
+                       RelativePermeabilityBase const & relperm )
+    :
+    m_phaseDens( fluid.phaseDensity() ),
+    m_dPhaseDens( fluid.dPhaseDensity() ),
+    m_phaseVisc( fluid.phaseViscosity() ),
+    m_dPhaseVisc( fluid.dPhaseViscosity() ),
+    m_phaseRelPerm( relperm.phaseRelPerm() ),
+    m_dPhaseRelPerm_dPhaseVolFrac( relperm.dPhaseRelPerm_dPhaseVolFraction() ),
+    m_phaseMob( subRegion.getField< fields::immiscibleMultiphaseFlow::phaseMobility >() ),
+    m_dPhaseMob( subRegion.getField< fields::immiscibleMultiphaseFlow::dPhaseMobility >() )
+  {}
+
+  /**
+   * @brief Performs the kernel launch
+   * @tparam POLICY the policy used in the RAJA kernels
+   * @tparam KERNEL_TYPE the kernel type
+   * @param[in] numElems the number of elements
+   * @param[inout] kernelComponent the kernel component providing access to the compute function
+   */
+  template< typename POLICY, typename KERNEL_TYPE >
+  static void
+  launch( localIndex const numElems,
+          KERNEL_TYPE const & kernelComponent )
+  {
+    forAll< POLICY >( numElems, [=] GEOS_HOST_DEVICE ( localIndex const ei )
+    {
+      kernelComponent.compute( ei );
+    } );
+  }
+
+  /**
+   * @brief Compute the phase mobilities in an element
+   * @tparam FUNC the type of the function that can be used to customize the kernel
+   * @param[in] ei the element index
+   * @param[in] phaseMobilityKernelOp the function used to customize the kernel
+   */
+  template< typename FUNC = NoOpFunc >
+  GEOS_HOST_DEVICE
+  void compute( localIndex const ei,
+                FUNC && phaseMobilityKernelOp = NoOpFunc{} ) const
+  {
+    using Deriv = immiscibleFlow::DerivativeOffset;
+
+    arraySlice1d< real64, immiscibleFlow::USD_PHASE - 1 > const phaseMob = m_phaseMob[ei];
+    arraySlice2d< real64, immiscibleFlow::USD_PHASE_DS - 1 > const dPhaseMob = m_dPhaseMob[ei];
+    int sign[2] = {1, -1};
```

Comment:
can you avoid using sign here and use fixed second index for `m_dPhaseRelPerm_dPhaseVolFrac` instead, like:
`real64 const dRelPerm_dS = m_dPhaseRelPerm_dPhaseVolFrac[ei][0][ip][0];`

Reply:
The mobility derivative matrix we are currently using is diagonal. This implementation handles missing off-diagonal entry.